docs/design/features/Apple-Silicon-UX.md

```diff
@@ -0,0 +1,107 @@
+
+### Other Observations
+
+- This mixed architecture problem is transient in nature. Essentially it means it will only exist for the support lifetime of the previous releases (2.1, 3.1, 5.0).  We want to avoid making major architectural changes to support this temporary design problem.
+
+- We want to keep the process parent-child relationships as consistent as possible.
+
+- We want minimal changes to the earlier releases.
+
+### Proposed Side-By-Side Design
+
+For Apple Silicon the proposed plan is to:
+
+- On Apple Silicon .NET runtime versions less than 6.0 will always require `x86_64` architecture. All other runtime versions will require `arm64` architecture on Apple Silicon.
+
+- Make the .NET 6 `dotnet` & `libhostfxr.dylib` (`hostfxr`) universal binaries. This means both native and emulated processes can perform runtime resolution to identify the correct runtime to use.
+
+- The `hostfxr` will be modified to enforce the architecture version  policy above. It will return an architecture mismatch error when the runtime (by version policy above) require a different architecture than the currently running process.
+
+- `dotnet` will be modified to enable switching the architecture of the current process when it sees an architecture mismatch error returned by `hostfxr`. The current process will switch architecture as needed and run with the same arguments as the new process. The switching will use `posix_spawn` with changes to `posix_spawnattr_t` to set `POSIX_SPAWN_SETEXEC` and preferred architecture via `posix_spawnattr_setbinpref_np` (see Apple Silicon man page) and setting `sysctl` `kern.curproc_arch_affinity` property appropriately.
+
+- The installer will be modified to enforce the .NET 6 architecture policy. It will prevent osx-x64 install on Apple Silicon
+
+- The existing installer releases will need to be tested to make sure they do not clobber the new `dotnet` modifications.
+
+#### Implications
+
+- The proposed version policy will prohibit us from making a native port of earlier runtime releases. .NET 2.1 will be out of support before .NET 6.0 releases. .NET 5.0 will be out of support 3 month after release of .NET 6. .NET 3.1 is the only viable candidate for back porting Apple Silicon support.  However, given the size of the changes to the runtime and the JIT it is highly unlikely this is something we would actively pursue.
+
+- The proposed version policy will prevent .NET 6 apps published with rid of osx-x64 from running on Apple Silicon even with Rosetta 2 emulation. This is desirable because long term we want to stop supporting Rosetta 2 emulation on Apple Silicon. IT support grows our effective platform support matrix and increases our support burden. We are dependent on Apple to fix Rosetta 2 issues, but they are dependent on us to root cause them. Performance is also expected to be better with native code.
```

Comment:
I expect that we will continue supporting running x64 applications published as self-contained and private installs of the x64 runtime (dotnet-install script) on Apple Silicon hardware.

Reply:
It may be worth describing why these two options are not sufficient workaround.

Reply:
I am mostly describing what (I think) #48541 implemented with respect to x64. When `hostfxr` found a runtime>= .NET 6.0 on x64, it marked it as `CorehostMustRunAsArchArm64`.  I debated only doing that for the muxer, but I finally let the change affect the `apphost` too.  So I think this describes what I think would happen with that change.

To support what you suggest is more complex.  Not impossible.  It probably needs more discussion.  Framework dependent `apphost` is the most complex.  I wasn't planning on making `apphost` a universal binary as it is a lot more complicated.

Reply:
Why does apphost need to be a universal binary for what I am describing?

Reply:
My knowledge of `apphost` is a little dated so it might be inaccurate.  I am assuming a framework dependent `apphost` will need to resolve a runtime from the global path.  So if the runtime needs to run as a different architecture, we need to have a policy.  I was using a hard version cut of to enable that.

If we wanted side by side x64 and arm64 versions we need to install them such that they can coexist.  i.e. add the arch to the runtime install path.

Otherwise we need to at least note that the app cannot be run.

I didn't want to enable switching the architecture of a published app, which would have required the universal `apphost`.  It would also be problematic for apps with native binaries which were not universal.

So allowing .NET 6 x64 apphost framework dependent apps to run on Apple Silicon is problematic.  Disabling for all .NET 6 seemed to be a simpler message.

Reply:
The scenario that I have mind is: application published for x64, with x64 apphost, x64 runtime installed using `dotnet-install`, and `DOTNET_ROOT` pointing to the place where the x64 private runtime is installed. I think this should continue working fine as it does today.

The scenario that you are describing is an application published for both `arm64` and `x64`, with universal apphost. I agree that it is more complex and we can start without it.

> It would also be problematic for apps with native binaries which were not universal.

Right, it would require SDK changes to binplace either universal or both x64 and native binaries into the published app.

Reply:
Apparently there are critical scenarios where we need to bridge this even for .NET 6.

> DOTNET_ROOT pointing to the place where the x64 private runtime is installed

The issue is, there is no way to tell whether `DOTNET_ROOT` points to `x64` runtime or `arm64` runtimes.  So if we enable this scenario for .NET 6, we need a mechanism or policy to determine the architecture of the runtime.  

In the e-mail discussion this was out of scope.  So I was assuming the simplest solution would be to detect using version and policy.  And just disable x64 runtime on arm64.

I'll update the draft to discuss the options here.